src/parser/mod.rs

```diff
@@ -0,0 +1,695 @@
+		Validator::new_with_features(features)
+			.validate_all(&self.bytes())
+			.map(|_| ())
+			.map_err(|err| anyhow!(err))
+	}
+
+	/// Registers a new raw_section in the ModuleInfo
+	pub fn section(&mut self, id: u8, range: Range<usize>, full_wasm: &[u8]) {
+		self.raw_sections.insert(id, RawSection::new(id, full_wasm[range].to_vec()));
+	}
+
+	/// Returns the function type based on the index of the type
+	/// `types[idx]`
+	pub fn get_type_by_idx(&self, type_idx: u32) -> Result<&Type> {
+		if type_idx >= self.types_map.len() as u32 {
+			return Err(anyhow!("type {} does not exist", type_idx));
+		}
+		Ok(&self.types_map[type_idx as usize])
+	}
+
+	/// Returns the function type based on the index of the function
+	/// `types[functions[idx]]`
+	pub fn get_type_by_func_idx(&self, func_idx: u32) -> Result<&Type> {
+		if func_idx >= self.function_map.len() as u32 {
+			return Err(anyhow!("function {} does not exist", func_idx));
+		}
+		let type_idx = self.function_map[func_idx as usize];
+		self.get_type_by_idx(type_idx)
+	}
+
+	pub fn resolve_type_idx(&self, t: &Type) -> Option<u32> {
+		let Type::Func(dt) = t else { todo!() };
+		for (index, ty) in self.types_map.iter().enumerate() {
+			let Type::Func(ot) = ty else { todo!() };
+			if ot.eq(dt) {
+				return Some(index as u32);
+			}
+		}
+		None
+	}
+
+	pub fn add_func_type(&mut self, func_type: &Type) -> Result<u32> {
+		let func_type_index = match self.resolve_type_idx(func_type) {
+			None => self.types_map.len() as u32,
+			Some(index) => return Ok(index),
+		};
+		// Add new type
+		let mut type_builder = wasm_encoder::TypeSection::new();
+		for t in &self.types_map {
+			DefaultTranslator.translate_type_def(t.clone(), &mut type_builder)?;
+		}
+		self.types_map.push(func_type.clone());
+		DefaultTranslator.translate_type_def(func_type.clone(), &mut type_builder)?;
+		self.replace_section(SectionId::Type.into(), &type_builder)?;
+		Ok(func_type_index)
+	}
+
+	/// Returns the number of globals used by the Wasm binary including imported
+	/// globals
+	#[allow(dead_code)]
+	pub fn get_global_count(&self) -> usize {
+		self.global_types.len()
+	}
+
+	/// Replace the `i`th section in this module with the given new section.
+	pub fn replace_section(
+		&mut self,
+		sec_type: u8,
+		new_section: &impl wasm_encoder::Section,
+	) -> Result<()> {
+		self.raw_sections
+			.insert(sec_type, RawSection::new(sec_type, truncate_len_from_encoder(new_section)?));
+		Ok(())
+	}
+
+	pub fn add_export(&mut self, name: &str, kind: ExportKind, index: u32) -> Result<()> {
+		let mut section_builder = wasm_encoder::ExportSection::new();
+
+		for export in self.export_section()? {
+			let export_kind = DefaultTranslator.translate_export_kind(export.kind).unwrap();
+			section_builder.export(export.name, export_kind, export.index);
+		}
+
+		section_builder.export(name, kind, index);
+		if let ExportKind::Global = kind {
+			self.exports_global_count += 1;
```

Comment:
How do you know if the name is not defined already? Overwriting existing export would affect other  stats in the `ModuleInfo`?

Reply:
Good point. I should've checked if the name is not duplicated.
Overwriting would not `ModuleInfo` stats if the `ExternalKind` is the same, otherwise I would need to make sure that self.exports_global_count is handled accordingly.